### PR TITLE
Add query for toilet availability on rest areas

### LINKED ISSUE #940: Rest area and highway=services have toilettes?
### General
Affected tag(s) to be modified/added: [](https://wiki.openstreetmap.org/wiki/Key:toilets)
Question asked: **Is there a toilet on the rest area?** yes/no

### Checklist
Checklist for quest suggestions (see guidelines):
- [x] 🚧 To be added tag is established and has a useful purpose
- [x] 🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
- [x] 🐿️ Easily answerable by everyone from the outside but a survey is necessary 
- [x] 💤 Not an overwhelming percentage of elements have the same answer (No spam)
- [x] 🕓 Applies to a reasonable number of elements (Worth the effort)

Perhaps we want to exclude areas where a toilet is mapped explict inside the area.

### PR DESCRIPTION
**Pull request by ENT8R**
This PR fixes #940 and adds a little bit more complexity into the toilet availability quest. I'm not sure if I made everything 100% right but this is what a PR is for... 😄

### CODE REVIEW

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java [hunk 1]
```diff
@@ -36,10 +63,89 @@ public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
 		changes.add("toilets", yesno);
 	}
 
+	@Nullable @Override public Boolean isApplicableTo(Element element) { return null; }
+
+	@Override public boolean download(BoundingBox bbox, final MapDataWithGeometryHandler handler)
+	{
+		List<AddHousenumber.ElementWithGeometry> items = downloadElementsWithoutToilets(bbox);
```
Comment (HolgerJeromin) [inline app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: Wrong name in the long run :)

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java [hunk 2]
```diff
@@ -36,10 +63,89 @@ public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
...
+		return success ? coords : null;
+	}
+
 	@Override public String getCommitMessage() { return "Add toilet availability"; }
 	@Override public int getIcon() { return R.drawable.ic_quest_toilets; }
+	@Override public int getTitle() { return R.string.quest_toiletAvailability_name_title; }
 	@Override public int getTitle(@NonNull Map<String, String> tags)
 	{
-		return R.string.quest_toiletAvailability_name_title;
+		boolean isRestArea = "rest_area".equals(tags.get("highway"));
```
Comment (HolgerJeromin) [inline app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: please add the bigger =services, too

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java [hunk 3]
```diff
@@ -2,27 +2,54 @@
...
+	private static String getElementsWithoutToiletsOverpassQuery(BoundingBox bbox)
 	{
 		// only for malls and big stores because users should not need to go inside a non-public
 		// place to solve the quest. (Considering malls and department stores public enough)
-		return "nodes, ways with shop ~ mall|department_store and name and !toilets";
+		return new FiltersParser().parse("nodes, ways with (shop ~ mall|department_store and name and !toilets) or (highway=rest_area and !toilets)").toOverpassQLString(bbox);
+	}
+
+	/** Query that returns all toilets */
+	private static String getFreeFloatingToiletsOverpassQuery(BoundingBox bbox)
```
Comment (HolgerJeromin) [inline app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: is it possible to filter toilets inside in overpass api?

Comment (ENT8R) [reply app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: Not that I know of... It is done the same with the housenumber quest like with this quest...

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java (hunk 2 above)
Comment (ENT8R) [reply app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: Ah... I wasn't aware of them. I will add them too

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java (hunk 1 above)
Comment (ENT8R) [reply app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: Do you have another suggestion?

**Comment by ENT8R** [comment]
After thinking a little bit about this PR, I don't know whether this PR makes sense at all or just adds too much complexity... I think that one can drop the whole logic with "don't show quest if there is already a node inside the element" because IMO it does make sense too to add this information if a separately mapped toilet does already exist. It is also valuable information and not really duplicate like it would be with house numbers...
Are there any other opinions? If no I will close this and make a new PR to add only the necessary functionality.

**Comment by HolgerJeromin** [comment]
The reason to exclude them would be that this can be extracted in the armchair and we should not ask calculateable data.
But in a rest areas there are not much quests so it is perhaps not super important

File: app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java (hunk 1 above)
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/quests/toilet_availability/AddToiletAvailability.java]: > Do you have another suggestion?

Make the class not a inner class of AddHousenumber.

**Comment by ENT8R** [comment]
> For toilets,... pfff, they are not really that numerous anyway, so I find it okay if the app asks therefore for an information that is already there in another form.

I agree. So then we should rather just include ways with `highway=rest_area|services` without adding the complexity?

**Pull request closed without merging**